base/dates/io.jl

```diff
@@ -34,123 +34,153 @@ const MONTHTOVALUE = Dict{UTF8String,Dict{UTF8String,Int}}("english"=>english)
 const MONTHTOVALUEABBR = Dict{UTF8String,Dict{UTF8String,Int}}("english"=>abbrenglish)
 
 # Date/DateTime Parsing
-abstract Slot{P<:AbstractTime}
+abstract Slot{T<:Any}
 
-immutable DelimitedSlot{P<:AbstractTime} <: Slot{P}
-    i::Int
-    period::Type{P}
+immutable DelimitedSlot{T<:Any} <: Slot{T}
+    parser::Type{T}
+    letter::Char
     width::Int
-    option::Int
-    locale::AbstractString
+    transition::Union{Regex,AbstractString}
 end
 
-immutable FixedWidthSlot{P<:AbstractTime} <: Slot{P}
-    i::Int
-    period::Type{P}
+immutable FixedWidthSlot{T<:Any} <: Slot{T}
+    parser::Type{T}
+    letter::Char
     width::Int
-    option::Int
-    locale::AbstractString
 end
 
 immutable DateFormat
     slots::Array{Slot,1}
-    begtran # optional transition from the start of a string to the 1st slot
-    trans #trans[i] == how to transition FROM slots[i] TO slots[i+1]
+    prefix::AbstractString # optional transition from the start of a string to the 1st slot
+    locale::AbstractString
 end
 
-immutable DayOfWeekSlot <: AbstractTime end
+abstract DayOfWeekSlot
 
-duplicates(slots) = any(map(x->count(y->x.period==y.period,slots),slots) .> 1)
+# Slot rules allowed for where ismatch(r"^[a-zA-Z]$",c)
+slotrule(::Type{Val{'y'}}) = Year
+slotrule(::Type{Val{'m'}}) = Month
+slotrule(::Type{Val{'u'}}) = Month
+slotrule(::Type{Val{'U'}}) = Month
+slotrule(::Type{Val{'e'}}) = DayOfWeekSlot
+slotrule(::Type{Val{'E'}}) = DayOfWeekSlot
+slotrule(::Type{Val{'d'}}) = Day
+slotrule(::Type{Val{'H'}}) = Hour
+slotrule(::Type{Val{'M'}}) = Minute
+slotrule(::Type{Val{'S'}}) = Second
+slotrule(::Type{Val{'s'}}) = Millisecond
+slotrule{c}(::Type{Val{c}}) = Union{}
```

Comment:
I originally had implemented this as a `Dict` but switched to using value types as made the code appear intended to be extensible rather than just happen to be. The one advantage of switching to a `Dict` gets us is that the regular expression in `DateFormat` can restricted to the letters used.

I'm intending to stick with the dynamic dispatch solution unless someone has a solid reason against it.